Move permissions and permissions.* except .local to /usr/share/permissions

thkukuk opened the pull request:
/etc/permissions and /etc/permissions.* (except .local) are neither configuration files nor templates, an admin should never modify them.
Move this files to /usr/share/permissions where they belong according to the FHS.

maltek commented:
Thanks! This has been bugging me for a while. (And I just now noticed these files never even had a warning not to edit them like polkit-default-privs which makes it so much worse.)

There's a few things I'm a bit uncertain about:

1. You explicitly read `/etc/permissions` and `/etc/permissions.{easy,secure,paranoid}` if they exist. This isn't wrong at all, but opens a can of worms I'm not sure we want to deal with - next thing we know, we have to add ways to selectively override or remove entries from the pre-installed profile. Until now, editing these files in `/etc` was not supported at all - they would get just get silently overwritten on the next permissions update. So I think I'd rather get rid of these files in `etc` completely. (The `local` profile is the long-supported way to do the same thing.)
1. There's still a few places with hardcoded `/etc/` paths, at least lines 355 and 364 I think need updating. (Line 434 seems fine, since it's only about custom profiles.)

Before we can merge this, we'll have to take care of a few things:

 - update the whitelistings in rpmlint-tests to cover the new paths
 - check where the YaST UI gets the possible values for the profile from. If the list there is dynamic, they'll have to make this change first.

(After the changes are through, we'll also want to update the 3 packages that drop files into `/etc/permissions.d` to use `/usr/share` instead.)

thkukuk commented:
I keeped /etc/permissions and /etc/permissions.{easy,secure,paranoid} as fallback to allow a smoothless migration. chkstat and permissions-config are separate RPMs. Better keep this fallback then breaking systems.

Line 355 and 364 are bugs, thanks for catching that. I will fix that now.

We also need to take care of /etc/apparmor/severity.db and maybe more apparmor files.
Which rpmlint-tests do you mean? I don't see any warnings from rpmlint.

maltek commented:
> I keeped /etc/permissions and /etc/permissions.{easy,secure,paranoid} as fallback to allow a smoothless migration. chkstat and permissions-config are separate RPMs. Better keep this fallback then breaking systems.

That is a good point, but I'd rather deal with that by changing the RPM requirements to specify exact version numbers.


> Which rpmlint-tests do you mean? I don't see any warnings from rpmlint.

We have a hardcoded whitelist of packages in rpmlint-tests that are allowed to drop files into `/etc/permissions.d` so that we (security team) get to check the security of setuid binaries before they make it into the distro. When we add new paths for `chkstat`, the list of sensitive paths must be updated or this whitelisting mechanism becomes ineffective. I can take care of that tomorrow.

As for YaST, I just checked and it has a hardcoded list of profiles. Only some help texts need adjusting but that can wait.

thkukuk commented:
On Mon, Dec 09, maltek wrote:

> That is a good point, but I'd rather deal with that by changing the RPM
> requirements to specify exact version numbers.

This will not help at all, there is no way to make sure that this
packages will be installed in an atomic way.

thkukuk commented:
YaST only reads /etc/sysconfig/security, but none of the permissions files. Nothing we need to care about.

So there are only the other packages which write into /etc/permissions.d. As we don't know which external packages use this directory, too, we have to keep that anyways. I don't see how/why this should block this PR.

maltek commented:
> This will not help at all, there is no way to make sure that this
packages will be installed in an atomic way.

Huh? permissions-config already has a `Requires(post)` on chkstat. Assuming this directive actually works, then I don't see how the `chkstat --system` could fail even if we changed it to specify a specific version and added some additional version-specific `Requires` to the permissions package.

Anyway, I'd also be happy if `chkstat` would only fall back to the paths in `etc` when the path in `usr/share` does not exist. I do want to avoid supporting files installed at both locations at the same time.

> So there are only the other packages which write into /etc/permissions.d. As we don't know which external packages use this directory, too, we have to keep that anyways. I don't see how/why this should block this PR.

Nono, I don't want all packages that place files there to block the PR at all - after all, they can't move to the new directory until this PR has landed. I do want to update the whitelisting mechanism before this PR lands because otherwise our whitelisting would be ineffective for some time. (BTW, this whitelisting means we do know [exactly which packages](https://github.com/openSUSE/rpmlint-checks/blob/00e6393112de7c6da46780842cd787f693b05af3/CheckSUIDPermissions.py#L19-L32) place files there.)

thkukuk commented:
On Tue, Dec 10, maltek wrote:

> Huh? permissions-config already has a Requires(post) on chkstat.

Which only means that there is a chkstat version installed before
permissions-config will be installed, but this says nothing about
update at all (even if libzypp will try to honour this for updates,
too) and this does not mean that the system is functional in between. 
Other packages could be installed in between, which call a chckstat 
binary not finding his configuration files.

> Assuming this directive actually works, then I don't see how the 
> chkstat --system could fail even if we changed it to specify a 
> specific version and added some additional
> version-specific Requires to the permissions package.

Because the update is not done in an atomic way only with this two
packages. This is the biggest common mistake to assume and one of the 
most often seen reasons why updates go wrong/fail.

> Anyway, I'd also be happy if chkstat would only fall back to the paths in etc
> when the path in usr/share does not exist. I do want to avoid supporting files
> installed at both locations at the same time.

Should be easy to do.

> (BTW, this whitelisting means we do know exactly which packages place 
> files there.)

Don't make the mistake to assume that all RPMs our customers install 
will ever only be build in our build service with our checks. We have no 
idea what our customers are building, installing and from where they get 
it.

maltek commented:
FYI: The rpmlint changes are currently blocked by some other problems with the package in OBS. And I won't be in the office for the next 3 weeks to push this forward immediately after these are resolved.

thkukuk commented:
What is the status here?

thkukuk commented:
@maltek any update?

maltek commented:
Sorry, this is still blocked by these rpmlint build problems. They've been "nearly solved" for a week now, @jsegitz may know more.

jsegitz commented:
> Sorry, this is still blocked by these rpmlint build problems.

That might have been a misunderstanding. The problem I was working on didn't require changes to rpmlint itself, but to dependent packages. 

I just pulled the current git versions into rpmlint will submited to the devel project once I fixed the current problems with rpmlint-tests. Should land in Factory in the next couple of days.

mgerstner commented:
We can close this PR, the changes have been integrated via pull request #38.
